[Bug]: Bonjour discovery remains active after Gateway startup failure

### Bug type

Regression (worked before, now fails)

### Beta release blocker

No

### Summary

Gateway startup leaks its already-started Bonjour discovery advertiser when a later early-startup step fails before the discovery disposer reaches lifecycle ownership.

### Steps to reproduce

1. Start early Gateway runtime with a fake discovery service that returns a tracked async stop callback.
2. Make the subsequent remote-skills registry registration throw a specific startup error.
3. Run the existing startup-failure shutdown path and inspect the tracked discovery stop callback.

### Expected behavior

An acquired discovery disposer is owned immediately, startup-failure shutdown invokes it exactly once, and the original startup error remains the rejection even if discovery cleanup fails.

### Actual behavior

The original startup error survives, but the discovery stop callback is invoked zero times because it is returned and published only after all early runtime initialization succeeds. The advertiser/responder and installed console/process-error handlers remain active.

### OpenClaw version

Current main at 2753351abe63260f9f4a3381754c63fb5ed93b7d; also confirmed in stable v2026.6.34.

### Operating system

macOS; the lifecycle ownership defect is platform-independent.

### Install method

Source checkout with mocked, in-memory discovery boundaries.

### Model

Not applicable: failure occurs before model execution.

### Provider / routing chain

Not applicable: no provider or network service is used.

### Logs

```text
FAIL startGatewayEarlyRuntime > stops acquired discovery exactly once after later startup failure
AssertionError: expected "vi.fn()" to be called once, but got 0 times
```

### Impact and severity

Any Gateway startup failure after successful local discovery can leave Bonjour process resources and hooks installed. The leak occurs deterministically for the injected post-discovery failure.

### Additional information

The regression was introduced by #63975, which moved disposer publication after fallible early initialization. The aggregate Gateway discovery disposer is not idempotent, so resident shutdown and lifecycle shutdown must consume one authoritative owner exactly once.

--- a/src/gateway/server-core-runtime.ts
+++ b/src/gateway/server-core-runtime.ts
@@ -188,6 +188,7 @@ export async function startGatewayCoreRuntime(input: {
         log,
         logDiscovery,
         nodeRegistry,
+        swapBonjourStop: kernel.swapBonjourStop,
         pluginRegistry: pluginRuntime.registry,
         broadcast,
         nodeSendToAllSubscribed,
@@ -228,10 +229,7 @@ export async function startGatewayCoreRuntime(input: {
   const discoveryResident = residentRegistry.register({
     name: "bonjour-discovery",
     start: startEarlyRuntime,
-    stop: async () => {
-      const earlyRuntime = await startEarlyRuntime();
-      await earlyRuntime.bonjourStop?.();
-    },
+    stop: async () => await kernel.swapBonjourStop(null)?.(),
   });
   const taskAndSkillsResident = residentRegistry.register({
     name: "task-and-skills-runtime",

--- a/src/gateway/server-kernel.test.ts
+++ b/src/gateway/server-kernel.test.ts
@@ -20,7 +20,7 @@ import { createSyntheticPluginRuntimeClient } from "./server-plugin-runtime-clie
 
 describe("createGatewayKernel", () => {
   it("reports startup and readiness as draining during a direct close", async () => {
-    const port = await getFreePort();
+    const port = 19_789;
     const state = await createOpenClawTestState({
       label: "gateway-kernel-direct-close-readiness",
       layout: "home",
@@ -56,6 +56,20 @@ describe("createGatewayKernel", () => {
       expect(getStartup()).toMatchObject({ ok: true, status: "started" });
       expect(getReadiness()).toMatchObject({ ready: true, failing: [] });
 
+      const discoveryResident = kernel.residentRegistry
+        .list()
+        .find((resident) => resident.name === "bonjour-discovery");
+      if (!discoveryResident) {
+        throw new Error("Expected the Gateway discovery resident");
+      }
+      const residentFirstStop = vi.fn(async () => {});
+      kernel.kernel.swapBonjourStop(residentFirstStop);
+      await discoveryResident.stop();
+      expect(residentFirstStop).toHaveBeenCalledOnce();
+      expect(kernel.runtimeState.bonjourStop).toBeNull();
+
+      const closeFirstStop = vi.fn(async () => {});
+      kernel.kernel.swapBonjourStop(closeFirstStop);
       const configReloaderStop = createDeferred();
       vi.spyOn(kernel.runtimeState.configReloader, "stop").mockReturnValue(
         configReloaderStop.promise,
@@ -66,6 +80,9 @@ describe("createGatewayKernel", () => {
       expect(getReadiness()).toMatchObject({ ready: false, failing: ["gateway-draining"] });
       configReloaderStop.resolve();
       await closing;
+      await discoveryResident.stop();
+      expect(closeFirstStop).toHaveBeenCalledOnce();
+      expect(kernel.runtimeState.bonjourStop).toBeNull();
     } finally {
       try {
         await kernel?.closeOnStartupFailure();

--- a/src/gateway/server-lifecycle.ts
+++ b/src/gateway/server-lifecycle.ts
@@ -249,11 +249,9 @@ export async function prepareGatewayLifecycle(params: {
       runtimeState.gatewayMethods.splice(0, runtimeState.gatewayMethods.length, ...methods);
     },
     setEarlyRuntimeHandles: (handles: {
-      bonjourStop: typeof runtimeState.bonjourStop;
       getActiveTaskCount: () => number;
       skillsChangeUnsub: typeof runtimeState.skillsChangeUnsub;
     }) => {
-      runtimeState.bonjourStop = handles.bonjourStop;
       activeTaskCount.get = handles.getActiveTaskCount;
       runtimeState.skillsChangeUnsub = handles.skillsChangeUnsub;
     },
@@ -514,7 +512,7 @@ export async function prepareGatewayLifecycle(params: {
     const transport = transportBridge.current();
     await transport?.portalService.closeAll();
     await shutdownRuntime.createGatewayCloseHandler({
-      bonjourStop: runtimeState.bonjourStop,
+      bonjourStop: kernel.swapBonjourStop(null),
       tailscaleCleanup: runtimeState.tailscaleCleanup,
       clearSecretsRuntimeSnapshot: clearSecretsRuntimeSnapshotState,
       channelIds,

--- a/src/gateway/server-startup-early.test.ts
+++ b/src/gateway/server-startup-early.test.ts
@@ -2,6 +2,7 @@
  * Early gateway startup helper tests.
  */
 import { beforeEach, describe, expect, it, vi } from "vitest";
+import { runGatewayShutdownSteps } from "./server-shutdown.js";
 import { createGatewayMaintenanceStateForTest } from "./test-helpers.maintenance-state.js";
 
 type StartGatewayDiscovery = typeof import("./server-discovery-runtime.js").startGatewayDiscovery;
@@ -82,6 +83,7 @@ function earlyRuntimeInput(
     log,
     logDiscovery: log,
     nodeRegistry: {} as never,
+    swapBonjourStop: () => null,
     ...maintenanceState,
     skillsRefreshDelayMs: 30_000,
     getSkillsRefreshTimer: () => null,
@@ -150,6 +152,48 @@ describe("startGatewayEarlyRuntime", () => {
     expect(mocks.closeSkillsWatchers).toHaveBeenCalledTimes(1);
   });
 
+  it.each([false, true])(
+    "stops acquired discovery exactly once after later startup failure (cleanup rejects: %s)",
+    async (cleanupRejects) => {
+      const startupError = new Error("remote skills registry failed");
+      const cleanupError = new Error("discovery cleanup failed");
+      const stopDiscovery = vi.fn(async () => {
+        if (cleanupRejects) {
+          throw cleanupError;
+        }
+      });
+      const owner: { current: (() => Promise<void>) | null } = { current: null };
+      const swapBonjourStop = (next: typeof owner.current) => {
+        const previous = owner.current;
+        owner.current = next;
+        return previous;
+      };
+      mocks.startGatewayDiscovery.mockResolvedValueOnce({ bonjourStop: stopDiscovery });
+      mocks.setSkillsRemoteRegistry.mockImplementationOnce(() => {
+        throw startupError;
+      });
+      const onCleanupError = vi.fn();
+
+      const startup = startGatewayEarlyRuntime(
+        earlyRuntimeInput({ minimalTestGateway: false, swapBonjourStop }),
+      ).catch(async (error: unknown) => {
+        await runGatewayShutdownSteps({
+          steps: [
+            { name: "discovery resident", run: async () => await swapBonjourStop(null)?.() },
+            { name: "gateway close", run: async () => await swapBonjourStop(null)?.() },
+          ],
+          onError: onCleanupError,
+        });
+        throw error;
+      });
+
+      await expect(startup).rejects.toBe(startupError);
+      expect(stopDiscovery).toHaveBeenCalledOnce();
+      expect(owner.current).toBeNull();
+      expect(onCleanupError).toHaveBeenCalledTimes(cleanupRejects ? 1 : 0);
+    },
+  );
+
   it("broadcasts remote-node skill invalidations to operator clients", async () => {
     const broadcast = vi.fn();
 
@@ -212,6 +256,9 @@ describe("startGatewayEarlyRuntime", () => {
   });
 
   it("fails before discovery and task maintenance when task state cannot restore", async () => {
+    const stopDiscovery = vi.fn(async () => {});
+    const swapBonjourStop = vi.fn(() => null);
+    mocks.startGatewayDiscovery.mockResolvedValue({ bonjourStop: stopDiscovery });
     mocks.ensureTaskRuntimeStateReady.mockImplementationOnce(() => {
       throw new Error("task-flow registry restore failed");
     });
@@ -220,11 +267,14 @@ describe("startGatewayEarlyRuntime", () => {
       startGatewayEarlyRuntime(
         earlyRuntimeInput({
           minimalTestGateway: false,
+          swapBonjourStop,
         }),
       ),
     ).rejects.toThrow("task-flow registry restore failed");
 
     expect(mocks.startGatewayDiscovery).not.toHaveBeenCalled();
+    expect(swapBonjourStop).not.toHaveBeenCalled();
+    expect(stopDiscovery).not.toHaveBeenCalled();
     expect(mocks.configureTaskRegistryMaintenance).not.toHaveBeenCalled();
     expect(mocks.startTaskRegistryMaintenance).not.toHaveBeenCalled();
   });

--- a/src/gateway/server-startup-early.ts
+++ b/src/gateway/server-startup-early.ts
@@ -71,6 +71,7 @@ export async function startGatewayEarlyRuntime(params: {
     warn: (msg: string) => void;
   };
   nodeRegistry: Parameters<typeof import("../skills/runtime/remote.js").setSkillsRemoteRegistry>[0];
+  swapBonjourStop: (next: (() => Promise<void>) | null) => (() => Promise<void>) | null;
   pluginRegistry?: PluginRegistry;
   broadcast: GatewayMaintenanceParams["broadcast"];
   nodeSendToAllSubscribed: Parameters<StartGatewayMaintenanceTimers>[0]["nodeSendToAllSubscribed"];
@@ -102,8 +103,11 @@ export async function startGatewayEarlyRuntime(params: {
       ensureTaskRuntimeStateReady();
     });
   }
-  const bonjourStop = await measureStartup(params.startupTrace, "runtime.early.discovery", () =>
-    startGatewayPluginDiscovery(params),
+  // Startup failure can occur immediately after discovery; publish its owner first.
+  params.swapBonjourStop(
+    await measureStartup(params.startupTrace, "runtime.early.discovery", () =>
+      startGatewayPluginDiscovery(params),
+    ),
   );
   let getActiveTaskCount = () => 0;
 
@@ -205,7 +209,6 @@ export async function startGatewayEarlyRuntime(params: {
   };
 
   return {
-    bonjourStop,
     getActiveTaskCount,
     skillsChangeUnsub,
     startMaintenance,

